### PR TITLE
eint and unit tests

### PR DESCRIPTION
**Pull request by mikurys**
eint_minus and eint_plus for integral types, make tab_view read only,  with_mutex fix and unit tests

### CODE REVIEW

File: src/stdplus/eint.hpp [hunk 1]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
+	static_assert(std::numeric_limits<T>::min()<=0, "Integral type not include 0");
+	static_assert(std::numeric_limits<T>::max()>=0, "Integral type not include 0"); // it is normal integer with 0 value
+	bool is_overflow=false;
+	if (a>0)
+	{
+		if (std::numeric_limits<T>::max()-a < b) is_overflow = true;
+	}
+	if (a<0)
+	{
+		if (std::numeric_limits<T>::min()-a > b) is_overflow = true;
```
Comment (robertoleksy) [inline src/stdplus/eint.hpp]: it will be works only for unsigned integers [basic.fundamental] 3.9.1/4, for signed this is UB [expr] 5/4

File: src/stdplus/eint.hpp [hunk 2]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
-T eint_plus(T a, T b) {
-	T c = a + b;
-	return c;
+T eint_plus(T a, T b, typename std::enable_if<std::is_integral<T>::value>::type* = 0) {
+	static_assert(std::numeric_limits<T>::min()<=0, "Integral type not include 0");
+	static_assert(std::numeric_limits<T>::max()>=0, "Integral type not include 0"); // it is normal integer with 0 value
+	bool is_overflow=false;
+	if (a>0)
+	{
+		if (std::numeric_limits<T>::max()-a < b) is_overflow = true;
```
Comment (robertoleksy) [inline src/stdplus/eint.hpp]: min?

File: src/stdplus/tab.hpp [hunk 3]
```diff
@@ -148,14 +148,19 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
 		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
 		tab_range(TC & obj, size_t begin, size_t end)
 				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
-			if (std::less<const T*>()(end,begin)) throw std::invalid_argument("Incorect begin or end (end<begin).");
+			if (end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (robertoleksy) [inline src/stdplus/tab.hpp]: why not std::less?

File: src/stdplus/eint.hpp (hunk 2 above)
Comment (mikurys) [inline src/stdplus/eint.hpp]: possible overflows:
if a is positive, always a+b > min (we assume min<=0, min <= a,b <= max  like fundamental types) so we only have to check condition
a+b > max  ===> a > max-b ===> b > max -a

File: src/stdplus/tab.hpp (hunk 3 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: end, begin are size_t so we can use operator< too

File: src/stdplus/tab.hpp [hunk 4]
```diff
@@ -148,14 +148,19 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
 		}
 
 		T & at(size_t ix) {
 			if (ix >= size() || ix < 0) throw std::out_of_range("Reading tab_range out of range");
 			return *( m_begin + ix );
 		}
 
+		const T & at(size_t ix) const {
+			if (ix >= size() || ix < 0) throw std::out_of_range("Reading tab_view out of range");
```
Comment (rfree-d) [inline src/stdplus/tab.hpp]: instead of testing   ix<0   do e.g.
static_assert( std::limits< typedecl(ix) > min ......                  >= 0);  // no need to test ix<0 since it's type is non-negative

and ditto in other such places

File: src/stdplus/tab.hpp [hunk 5]
```diff
@@ -200,34 +203,22 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
...
 
-		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
-		tab_view(TC & obj, size_t begin, size_t end)
-				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
-			if(end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
-		}
-
-		T & at(size_t ix) {
+		const T & at(size_t ix) const {
 			if (ix >= size() || ix < 0) throw std::out_of_range("Reading tab_view out of range");
```
Comment (rfree-d) [inline src/stdplus/tab.hpp]: ditto ix<0

File: src/stdplus/vector_mutexed_obj.hpp [hunk 6]
```diff
@@ -97,15 +101,16 @@ void vector_mutexed_obj<TObj>::push_back(std::unique_ptr<TObj> && obj_ptr) {
...
 	{
 		UniqueLockGuardRO<MutexShared> lg_all(m_mutex_all);
 		one_object_with_mutex = m_data.at(ix).get();
 		// unlocking container
 	}
 
 	{
-		auto lg_one( one_object_with_mutex->get_lock_RW() );
+		auto &mutex = one_object_with_mutex->get_mutex();
+		UniqueLockGuardRW<MutexShared> lg_one(mutex);
```
Comment (rfree-d) [inline src/stdplus/vector_mutexed_obj.hpp]: not fixed in my newloop?

better use one-line without mutex&

File: src/test/with_mutex.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		UniqueLockGuardRO<MutexShared> lg_ro(mutex);
+		EXPECT_EQ(obj2.get(lg_ro), value);
+	}
+	obj = std::move(2);
+	{
+		auto &mutex = obj.get_mutex();
+		UniqueLockGuardRO<MutexShared> lg_ro(mutex);
+		EXPECT_EQ(obj.get(lg_ro), 2);
+	}
+}
```
Comment (rfree-d) [inline src/test/with_mutex.cpp]: could use tests for:

use_RW()
use_RO()

were added a bit later

File: src/cable/asio_ioservice_manager.cpp [hunk 8]
```diff
@@ -87,7 +87,7 @@ boost::asio::io_service &c_asioservice_manager::get_next_ioservice() {
 	guard_inv;
 
 	m_last_ioservice++;
-	m_last_ioservice %= m_ioservice_array.size();
+	m_last_ioservice %= size();
```
Comment (rfree-d) [inline src/cable/asio_ioservice_manager.cpp]: indeed my error, thanks

:cold_sweat:

File: src/stdplus/vector_mutexed_obj.hpp (hunk 6 above)
Comment (mikurys) [inline src/stdplus/vector_mutexed_obj.hpp]: it is only test

File: src/stdplus/eint.hpp [hunk 9]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
 /// ...check if b can be properly expanded to T1 ?
 template <typename T>
-T eint_minus(T a, T b) {
-	if (a<b) {
+T eint_minus(T a, T b, typename std::enable_if<std::is_integral<T>::value>::type* = 0) {
+	static_assert(std::numeric_limits<T>::min()<=0, "Integral type not include 0");
+	static_assert(std::numeric_limits<T>::max()>=0, "Integral type not include 0"); // it is normal integer with 0 value
+	bool is_overflow=false;
+	if (b<0)
+	{
```
Comment (mikurys) [inline src/stdplus/eint.hpp]: Maybe because this file has many authors.
Fixed.

File: src/stdplus/eint.hpp [hunk 10]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
 
 /// Calculate (a-b) in safe way (see documentation and WARNINGS about this namespace)
-/// @TODO now assumes T is UNSIGNED type!
 /// @TODO can variable "b" be silently converted to type T here, including narrowing? if yes then template on T1,T2 and
 /// ...check if b can be properly expanded to T1 ?
 template <typename T>
-T eint_minus(T a, T b) {
-	if (a<b) {
+T eint_minus(T a, T b, typename std::enable_if<std::is_integral<T>::value>::type* = 0) {
+	static_assert(std::numeric_limits<T>::min()<=0, "Integral type not include 0");
```
Comment (mikurys) [inline src/stdplus/eint.hpp]: fixed

File: src/stdplus/eint.hpp [hunk 11]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
+	static_assert(std::numeric_limits<T>::max()>=0, "Integral type not include 0"); // it is normal integer with 0 value
+	bool is_overflow=false;
+	if (b<0)
+	{
+		if (std::numeric_limits<T>::max()+b < a) is_overflow = true;
```
Comment (mikurys) [inline src/stdplus/eint.hpp]: ok

File: src/stdplus/eint.hpp [hunk 12]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
+	}
+	if (b>0)
+	{
+		if (std::numeric_limits<T>::min()+b > a) is_overflow = true;
+	}
+	if (is_overflow) {
 		std::ostringstream oss; oss<<"Math error when calculating (" << a << " - " << b << ") for T="<<typeid(T).name();
 		throw std::range_error(oss.str());
```
Comment (mikurys) [inline src/stdplus/eint.hpp]: ok

File: src/stdplus/eint.hpp [hunk 13]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
...
+	static_assert(std::numeric_limits<T>::max()>=0, "Integral type not include 0"); // it is normal integer with 0 value
+	bool is_overflow=false;
+	if (a>0)
+	{
+		if (std::numeric_limits<T>::max()-a < b) is_overflow = true;
+	}
+	if (a<0)
+	{
+		if (std::numeric_limits<T>::min()-a > b) is_overflow = true;
+	}
```
Comment (mikurys) [inline src/stdplus/eint.hpp]: ok

File: src/stdplus/tab.hpp (hunk 3 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: ok

File: src/stdplus/tab.hpp (hunk 4 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: ok

File: src/stdplus/tab.hpp (hunk 5 above)
Comment (mikurys) [inline src/stdplus/tab.hpp]: ok

File: src/stdplus/eint.hpp [hunk 14]
```diff
@@ -18,24 +19,49 @@ namespace stdplus {
 namespace eint {
 
 /// Calculate (a-b) in safe way (see documentation and WARNINGS about this namespace)
-/// @TODO now assumes T is UNSIGNED type!
 /// @TODO can variable "b" be silently converted to type T here, including narrowing? if yes then template on T1,T2 and
 /// ...check if b can be properly expanded to T1 ?
 template <typename T>
-T eint_minus(T a, T b) {
-	if (a<b) {
+T eint_minus(T a, T b, typename std::enable_if<std::is_integral<T>::value>::type* = 0) {
```
Comment (mikurys) [inline src/stdplus/eint.hpp]: ok

File: src/test/with_mutex.cpp (hunk 7 above)
Comment (mikurys) [inline src/test/with_mutex.cpp]: there are no use_RW and use_RO on yed/dev yet

File: src/stdplus/eint.hpp (hunk 11 above)
Comment (rfree-d) [inline src/stdplus/eint.hpp]: :ok_hand: also I mistaken should be 127 for the max() in the comment, that's fixed above :crystal_ball:

File: src/stdplus/tab.hpp [hunk 15]
```diff
@@ -148,16 +148,19 @@ typename T2::iterator copy_safe_apart(size_t n, const T1 & tab_src, T2 & tab_dst
 		template <typename TC> // TC = type of some container. it must be linear-memory. TODO: metaprogramming check
 		tab_range(TC & obj, size_t begin, size_t end)
 				: m_begin(&(obj.at(begin))), m_end(&(obj.at(end-1))+1), m_size(end-begin) {
-			if (end<begin) throw std::invalid_argument("Incorect begin or end (end<begin).");
+			// std::less not required for size_t, but using it to be like the other code above (and correct for review or copy-paste)
+			if (std::less<size_t>()(end,begin)) throw std::invalid_argument("Incorect begin or end (end<begin).");
```
Comment (rfree-d) [inline src/stdplus/tab.hpp]: typo "incorrect"
also in other places
@mikurys please fix at some point by the way
:pencil2:

**Comment by rfree-d** [comment]
@mikurys  other typo then "incorect" is "objects was found"  -> "objects were found"  :pen:
